### PR TITLE
precompile RIPEMD-160 + Tests

### LINKED ISSUE #359: feat: Precompile 0x03 RIPEMD-160
# Feature Request

**Describe the Feature Request**

feat: Explore / propose strategy (if applicable, MVP implementation) for EVM Precompiled contract -> 0x03 RIPEMD-160

https://www.evm.codes/precompiled

### PR DESCRIPTION
**Pull request by hurrikaanig**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

no precompile RIPEMD-160

Resolves #359 

## What is the new behavior?

precompile RIPEMD-160 :rocket:  

## Other information

Time spent on this PR:  1 day

### CODE REVIEW

File: tests/unit/src/kakarot/precompiles/test_ripemd160.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        disable_hint_validation=True,
+    )
+
+
+@pytest.mark.asyncio
+class TestEcAdd:
+    @pytest.mark.parametrize(
+        "calldata_len",
+        [128],
+        ids=["calldata_len128"],
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/precompiles/test_ripemd160.py:24]: remove this

**Review by ClementWalter — CHANGES_REQUESTED** [review]
can you add a test in `test_cases.py` ? otherwise we don't test that the right precompile is behind the right address.
You can copy/paste the test from the evm.playground

File: src/kakarot/precompiles/ripemd160.cairo [hunk 2]
```diff
@@ -0,0 +1,954 @@
...
+    let (a) = uint32_add(a, j_bcd);
+    let (a) = uint32_add(a, x);
+    let (a) = uint32_add(a, 0x50a28be6);
+    let (res1) = ROLASE(a, s, e);
+    let (res2) = ROL(c, 10);
+    return (res1=res1, res2=res2);
+}
+
+// This is taken from https://github.com/greenlucid/chess-cairo.
+func pow2(i : felt) -> (res : felt){
```
[the remark this answers is not in the document]
Comment (hurrikaanig) [reply src/kakarot/precompiles/ripemd160.cairo:697]: Sure I can do it

File: tests/integration/bytecode/test_cases.py [hunk 3]
```diff
@@ -1613,4 +1613,19 @@
         "id": "Precompiles - EC_RECOVER - playground test case",
         "marks": [pytest.mark.EC_RECOVER, pytest.mark.Precompiles],
     },
+    {
+        "params": {
+            "value": 0,
+            "code": "60ff600052602060206001601f600363fffffffffa50602051",
+            "calldata": "",
+            "stack": str(int("2c0c45d3ecab80fe060e5f1d7057cd2f8de5e557", 16)),
```
Comment (ClementWalter) [inline tests/integration/bytecode/test_cases.py:1621]: fyi, you can use
```
str(0x2c0c45d3ecab80fe060e5f1d7057cd2f8de5e557)
```

**Review by ClementWalter — APPROVED** [review]

**Comment by ClementWalter** [comment]
you need to rebase because the signature of the precompile has changed since https://github.com/sayajin-labs/kakarot/pull/404

File: tests/integration/bytecode/test_cases.py [hunk 4]
```diff
@@ -1618,7 +1618,7 @@
             "value": 0,
             "code": "60ff600052602060206001601f600363fffffffffa50602051",
             "calldata": "",
-            "stack": str(int("2c0c45d3ecab80fe060e5f1d7057cd2f8de5e557", 16)),
+            "stack": str("0x2c0c45d3ecab80fe060e5f1d7057cd2f8de5e557"),
```
Comment (ClementWalter) [inline tests/integration/bytecode/test_cases.py:1621]: I guess that your previous `str(int(..))` was `str(0x...)` but without the quotes. In python, you can type `0xf` directly and it prints out (it is a) `15`

File: tests/unit/src/kakarot/precompiles/test_ripemd160.py [hunk 5]
```diff
@@ -19,16 +20,25 @@ async def ripemd160(
...
-        ascii_msg = [ord(char) for char in msg]
-        expected_result_byte_array = list(bytes.fromhex(expected_result))
-        hash =  (await ripemd160.test__ripemd160(ascii_msg).call()).result[0]
-        assert hash == expected_result_byte_array
+    async def test_ripemd160_should_return_correct_hash(self, ripemd160, msg_len):
+        ripemd160_crypto = RIPEMD160.new()
+
+        # Build message to be hashed
+        message = "".join((chr(random.randint(0, 255))) for _ in range(msg_len))
+        message_bytes = bytes(message, encoding='utf8')
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/precompiles/test_ripemd160.py:31]: what about a `message_bytes = bytearray([random.randint(0,  255) for _ in range(len))`

File: tests/unit/src/kakarot/precompiles/test_ripemd160.py [hunk 6]
```diff
@@ -19,16 +20,25 @@ async def ripemd160(
...
+        "msg_len",
+            [random.randint(1, 300) for _ in range(3)]
     )
-    async def test_ripemd160_should_return_correct_hash(self, ripemd160, msg, expected_result):
-        ascii_msg = [ord(char) for char in msg]
-        expected_result_byte_array = list(bytes.fromhex(expected_result))
-        hash =  (await ripemd160.test__ripemd160(ascii_msg).call()).result[0]
-        assert hash == expected_result_byte_array
+    async def test_ripemd160_should_return_correct_hash(self, ripemd160, msg_len):
+        ripemd160_crypto = RIPEMD160.new()
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/precompiles/test_ripemd160.py:27]: could you just put this right before it's used for readability?

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**